### PR TITLE
Introduce that evaluation metrics and loss are not always the same

### LINKED ISSUE #323: Find a way to introduce that the metrics and the objective function are not always the same thing in the model evaluation module
This used to be a FIXME in the markdown as reported in #319.

### PR DESCRIPTION
**Pull request by ArturoAmorQ**
Fixes #323.

To keep the idea of having this as extra material, it is hidden in the solution of the exercise.

Optionally we can also mention pinball loss, but I felt that we don't want to enter into the details of what different loss functions do.

### CODE REVIEW

File: python_scripts/metrics_sol_02.py [hunk 1]
```diff
@@ -91,3 +91,41 @@
 }
 scores = pd.DataFrame(scores)
 scores
+
+# %% [markdown] tags=["solution"]
+# In the Regression Metrics notebook we mentioned the concept of loss function,
+# which is the metric to be optimized when training a model. In the case of a
```
Comment (glemaitre) [inline python_scripts/metrics_sol_02.py:97]:
```suggestion
# which is the metric optimized when training a model. In the case of a
```

File: python_scripts/metrics_sol_02.py [hunk 2]
```diff
@@ -91,3 +91,41 @@
 }
 scores = pd.DataFrame(scores)
 scores
+
+# %% [markdown] tags=["solution"]
+# In the Regression Metrics notebook we mentioned the concept of loss function,
```
Comment (glemaitre) [inline python_scripts/metrics_sol_02.py:96]:
```suggestion
# In the Regression Metrics notebook, we introduced the concept of loss function,
```

File: python_scripts/metrics_sol_02.py [hunk 3]
```diff
@@ -91,3 +91,41 @@
...
+# which is the metric to be optimized when training a model. In the case of a
+# linear regression, the fitting process consists in minimizing the mean squared
```
Comment (glemaitre) [inline python_scripts/metrics_sol_02.py:98]: I find "linear regression" to be to vague here. I would instead refer to `LinearRegression` that is the scikit-learn estimator minimizing the MSE.

File: python_scripts/metrics_sol_02.py [hunk 4]
```diff
@@ -91,3 +91,41 @@
...
+# error (MSE). Some estimators, such as the `HistGradientBoostingRegressor`, can
+# use different loss functions, to be set using the `loss` hyperparameter.
+#
+# Notice that the evaluation metrics and the loss function are not necessarily
```
Comment (glemaitre) [inline python_scripts/metrics_sol_02.py:102]:
```suggestion
# Notice that the evaluation metrics and the loss functions are not necessarily
```

File: python_scripts/metrics_sol_02.py [hunk 5]
```diff
@@ -91,3 +91,41 @@
...
+# the same. Let's see an example:
```
Comment (glemaitre) [inline python_scripts/metrics_sol_02.py:103]: I don't know if it would be correct mentioning that we use a loss function as a proxy of the final evaluation metric that we want to optimize?

File: python_scripts/metrics_sol_02.py [hunk 6]
```diff
@@ -91,3 +91,41 @@
...
+for loss_func in loss_functions:
+    model = HistGradientBoostingRegressor(loss=loss_func)
+    cv_results = cross_validate(model, data, target, scoring=scoring)
+    mse = -cv_results["test_neg_mean_squared_error"]
+    mae = -cv_results["test_neg_mean_absolute_error"]
+    scores["loss"].append(loss_func)
+    scores["MSE"].append(f"{mse.mean():.1f} ± {mse.std():.1f}")
+    scores["MAE"].append(f"{mae.mean():.1f} ± {mae.std():.1f}")
+scores = pd.DataFrame(scores)
+scores.set_index("loss", inplace=True)
```
Comment (glemaitre) [inline python_scripts/metrics_sol_02.py:122]: Let's not use `inplace=True` since it could be deprecated in the future.

File: python_scripts/metrics_sol_02.py [hunk 7]
```diff
@@ -91,3 +91,41 @@
...
+# linear regression, the fitting process consists in minimizing the mean squared
+# error (MSE). Some estimators, such as the `HistGradientBoostingRegressor`, can
+# use different loss functions, to be set using the `loss` hyperparameter.
+#
+# Notice that the evaluation metrics and the loss function are not necessarily
+# the same. Let's see an example:
+
+# %%
+# solution
+from sklearn.ensemble import HistGradientBoostingRegressor
```
Comment (glemaitre) [inline python_scripts/metrics_sol_02.py:107]:
```suggestion
from collections import defaultdict
from sklearn.ensemble import HistGradientBoostingRegressor
```

File: python_scripts/metrics_sol_02.py [hunk 8]
```diff
@@ -91,3 +91,43 @@
...
+scores = pd.DataFrame(scores)
+scores.set_index("loss")
+
+# %% [markdown] tags=["solution"]
+# Even if the score distributions overlap due to the presence of outliers in the
+# dataset, it is true that the average MSE is lower when `loss="squared_error`,
+# whereas the average MAE is lower when `loss="absolute_error` as expected.
+# Indeed, the choice of a loss function is made depending on the evaluation
+# metric that we want to optimize for a given use case.
+#
```
Comment (ogrisel) [inline python_scripts/metrics_sol_02.py:131]: To address Guillaume remark above, we could probably insert a paragraph here along the lines:

> Note that it is not always possible to optimize directly some metrics.
> For instance, classification models are often evaluated using metrics
> computed on hard class predictions (e.g. accuracy or F1) rather
> than from the continuous values returned by the `predict_proba`
> or `decision_function` methods. Such hard class prediction metrics
> typically evolve in a discontinuous manner when progressively changing > the internal parameters of the model during training. Therefore they
> cannot be directly optimized by learning solvers based on gradient
> descent and the likes (the details are belong the scope of this MOOC).
> So instead classifiers are typically trained by optimizing a
> loss function computed from some continuous output of the model.
> For instance `LogisticRegression` minimizes the `log_loss` applied to
> the `predict_proba` output of the model. The `log_loss` acts
> as a "surrogate loss" for discontinuous classification metrics such as
> the accuracy: by minimizing the surrogate loss, we maximize their
> accuracy. However scikit-learn does not provide surrogate losses for
> all possible classification metrics.

**Review by ogrisel — APPROVED** [review]
Here are some suggestions. Other than that LGTM.

File: python_scripts/metrics_sol_02.py [hunk 9]
```diff
@@ -91,3 +91,62 @@
...
+# If you feel like going beyond the contents of this MOOC, you can try different
+# combinations of loss functions and evaluation metrics.
+#
+# Notice that there are some metrics that cannot be directly optimized by
+# optimizing a loss function. This is the case for metrics that evolve in a
+# discontinuous manner with respect to the internal parameters of the model, as
+# learning solvers based on gradient descent or similar optimizers require
+# continuity (the details are belong the scope of this MOOC).
```
Comment (glemaitre) [inline python_scripts/metrics_sol_02.py:139]:
```suggestion
# continuity (the details are beyond the scope of this MOOC).
```
